### DISCUSSION TITLE
can we use btrfs as  lower filesystem for overlayfs

### DISCUSSION THREAD

**Discussion by ningmingxiao** [moby/moby#47544] [Q&A] [answered]
### Description

There are two ways use btrfs
1.Use the BTRFS storage driver
ref:https://docs.docker.com/storage/storagedriver/btrfs-driver/
2.Use BTRFS as as lower filesystem for overlayfs
ref:https://btrfs.readthedocs.io/en/latest/Interoperability.html#overlayfs

**Comment by corhere** [comment, marked as answer]
> Since kernel 4.15 the btrfs filesystem can be used as lower filesystem for overlayfs (supporting the rename modes of exchange and whiteout).

Yes, if your kernel is new enough.

**Reply by ningmingxiao** [reply]
What are the differences between them? @corhere
how to chose?

**Reply by corhere** [reply]
The btrfs storage driver uses native btrfs features (subvolumes and snapshots) for deduplicated image storage. The overlay2 storage driver uses overlayfs for deduplicated image storage. Overlayfs is a virtual file system which persists the files in directory structures in a compatible backing file system such as ext4 or btrfs.

Use the overlay2 driver for new deployments, even if you are using btrfs as the backing filesystem. It's the only storage driver from which we are planning to provide a migration to containerd image storage when we deprecate and remove all image storage drivers (a.k.a. graph drivers) from dockerd.
